nx/lib/nx/defn/grad.ex

```diff
@@ -1343,9 +1334,77 @@ defmodule Nx.Defn.Grad do
 
   ## General helpers
 
-  defp unbroadcast(%{shape: shape} = x, res, %{shape: shape}), do: {x, res}
+  defp revectorize_args(args, ans) do
```

Comment:
Let's only apply this if args has more than one element and there are vectorized axes.


Also please test `x * sin(y)` where y is vectorized.